internal/controller/dpuoperatorconfig_controller.go

```diff
@@ -106,26 +114,97 @@ func (r *DpuOperatorConfigReconciler) Reconcile(ctx context.Context, req ctrl.Re
+	if err := r.ensureNetworkResourcesInjector(ctx, dpuOperatorConfig); err != nil {
 		logger.Error(err, "Failed to ensure Network Resources Injector is running")
+		reconcileErrors = append(reconcileErrors, componentError{component: "NetworkResourcesInjector", err: err})
+	}
+
+	// Update status based on reconciliation results
+	if err := r.updateStatus(ctx, dpuOperatorConfig, reconcileErrors); err != nil {
+		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
+// setCondition sets the specified condition on the DpuOperatorConfig status
+func (r *DpuOperatorConfigReconciler) setCondition(dpuOperatorConfig *configv1.DpuOperatorConfig, condition metav1.Condition) {
+	meta.SetStatusCondition(&dpuOperatorConfig.Status.Conditions, condition)
+}
+
+// setReadyCondition sets the Ready condition to True with the given reason and message
+func (r *DpuOperatorConfigReconciler) setReadyCondition(dpuOperatorConfig *configv1.DpuOperatorConfig, reason, message string) {
+	r.setCondition(dpuOperatorConfig, metav1.Condition{
+		Type:    "Ready",
+		Status:  metav1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+// setNotReadyCondition sets the Ready condition to False with the given reason and message
+func (r *DpuOperatorConfigReconciler) setNotReadyCondition(dpuOperatorConfig *configv1.DpuOperatorConfig, reason, message string) {
+	r.setCondition(dpuOperatorConfig, metav1.Condition{
+		Type:    "Ready",
+		Status:  metav1.ConditionFalse,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+// initializeStatus sets initial NotReady condition if no conditions exist
+func (r *DpuOperatorConfigReconciler) initializeStatus(ctx context.Context, dpuOperatorConfig *configv1.DpuOperatorConfig) error {
+	if len(dpuOperatorConfig.Status.Conditions) == 0 {
+		r.setNotReadyCondition(dpuOperatorConfig, "WaitingForReconcile", "Waiting for reconcile to start DPU Daemons")
+		return r.Status().Update(ctx, dpuOperatorConfig)
+	}
+	return nil
+}
+
+// updateStatus updates the status based on reconciliation results
+func (r *DpuOperatorConfigReconciler) updateStatus(ctx context.Context, dpuOperatorConfig *configv1.DpuOperatorConfig, reconcileErrors []componentError) error {
+	logger := log.FromContext(ctx)
+
+	if len(reconcileErrors) > 0 {
+		// Set NotReady condition with the first error, including component info in reason
+		firstError := reconcileErrors[0]
+		reason := fmt.Sprintf("%sError", firstError.component)
+		r.setNotReadyCondition(dpuOperatorConfig, reason, firstError.err.Error())
+		if updateErr := r.Status().Update(ctx, dpuOperatorConfig); updateErr != nil {
+			logger.Error(updateErr, "Failed to update status with error condition")
+		}
+		return firstError.err
+	}
+
+	// All components reconciled successfully - set Ready condition
+	r.setReadyCondition(dpuOperatorConfig, "AllComponentsReady", "DPU Daemons and Network Functions are ready")
+	if updateErr := r.Status().Update(ctx, dpuOperatorConfig); updateErr != nil {
```

Comment:
I think we should "get" before update. Or use the mutation design. I think tracing the code you do multiple updates to the CR. The reason why this works is because the data structure is updated when "update" is called. But if "dpuOperatorConfig" is ever updated by a daemon then this breaks. You have a mutual exclusion issue.
Which is when you mentioned about one field belonging to the daemon and 1 field belonging to the controller. That it is not as easy as we picture since "update" is not really a merge and if it detects changes that it has not included it will fail.

Reply:
Get is on this line https://github.com/openshift/dpu-operator/pull/553/files/a04da8f6b5d746f93b26c861a7d223a25e021a59#diff-ac97373df236b7f27a14df2d71cf3b436098b60ff6ae8fccdb17c6a6266a33ecR108

If we would do a get just before the update, then that would only narrow the distance between the get and the update.

The daemon doesn't touch this. If it would, the update here would fail because resourceVersion would not be the same and we would error out on line https://github.com/openshift/dpu-operator/pull/553/files/a04da8f6b5d746f93b26c861a7d223a25e021a59#diff-ac97373df236b7f27a14df2d71cf3b436098b60ff6ae8fccdb17c6a6266a33ecR143. That's how k8s resolves the conflict (which doesn't exist here now, and probably will not exist ever for DpuOperatorConfig CR).

Reply:
As long as the daemons never touch the DPU Config CR, this should be good.

However this is a note for when we update the DPU CR modifications you plan to do.